File: contest/leetcode/count-primes.cpp
/* coding:utf-8
 * Copyright (C) dirlt
 */

#include <cmath>
#include <vector>
using namespace std;
class Solution {
   public:
    int countPrimes(int n) {
        if (n == 0) return 0;
        n -= 1;
        if (n <= 1) return 0;
        if (n == 2) return 1;
        vector<char> primes(n + 1);
        int upper = lround(sqrt(n)) + 2;
        for (int i = 2; i <= upper; i++) {
            int upper2 = n / i;
            for (int j = i; j <= upper2; j++) {
                primes[i * j] = 1;
            }
        }
        int ans = 1;
        for (int i = 3; i <= n; i++) {
            if (primes[i] == 0) {
                ans += 1;
            }
        }
        return ans;
    }
};

File: contest/leetcode/count-primes.py
#!/usr/bin/env python
# coding:utf-8
# Copyright (C) dirlt

class Solution:
    def countPrimes(self, n):
        """
        :type n: int
        :rtype: int
        """

        if n == 0: return 0
        n -= 1
        if n <= 1: return 0
        if n == 2: return 1

        primes = [1] * (n + 1)
        for i in range(2, int(round(n ** 0.5)) + 2):
            for j in range(i, n // i + 1):
                primes[i * j] = 0

        count = 1
        for i in range(3, n + 1):
            if primes[i] == 1:
                count += 1
        return count


if __name__ == '__main__':
    sol = Solution()
    print(sol.countPrimes(1500000))
